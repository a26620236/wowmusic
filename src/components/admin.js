import React from 'react';
import ReactDOM from 'react-dom'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";
import { db, firebase, storage } from '../static/js/firebase'

class Admin extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      playlist__name:'',
      playlist__category:'',
      playlist__desc:'',
      photo__file:'',
      select__playlist:'',
      song__album:'',
      song__name:'',
      song__singer:'',
      song__length:'',
      song__file:'',
    }
  }
  render() {
    return (
      <div className='admin'>
        <div className='albums'>
          <div className='list-header'>新增專輯</div>
          <div className='list-body'>
            <div className='wrapper'>
              <div className='album-inform'>
                <div className='inputbox'>
                  歌單名稱:
                  <input data-field='playlist__name' onChange={this.handleInput.bind(this)}></input>
                </div>
                <div className='inputbox'>
                  分類:
                  <input data-field='playlist__category' onChange={this.handleInput.bind(this)}></input>
                </div>
                <div className='inputbox'>
                  介紹:
                  <input data-field='playlist__desc' onChange={this.handleInput.bind(this)}></input>
                </div>
                <div className='inputbox'>
                  封面相片:
                  <input type='file' data-field='photo__file' onChange={this.handleInputFile.bind(this)}></input>
                </div>
                <div className='submit' onClick={this.addAlbum.bind(this)}>送出</div>
              </div>
            </div>
          </div>
        </div>
        <div className='songs'>
          <div className='list-header'>新增歌曲</div>
          <div className='list-body'>
            <div className='inputbox'>
              歌單名稱:
              <input data-field='select__playlist' onChange={this.handleInput.bind(this)}></input>
            </div>
            <div className='inputbox'>
              歌曲一:
              <input data-field='song__name' onChange={this.handleInput.bind(this)}></input>
            </div>
            <div className='inputbox'>
              演唱者:
              <input data-field='song__singer' onChange={this.handleInput.bind(this)}></input>
            </div>
            <div className='inputbox'>
              專輯名稱:
              <input data-field='song__album' onChange={this.handleInput.bind(this)}></input>
            </div>
            <div className='inputbox'>
              歌曲長度:
              <input data-field='song__length' onChange={this.handleInput.bind(this)}></input>
            </div>
            <div className='inputbox'>
              檔案:
              <input type='file' data-field='song__file' onChange={this.handleInputFile.bind(this)} ></input>
            </div>
            <div className='submit' onClick={this.addSong.bind(this)}>送出</div>
          </div>
        </div>
      </div>
    )
  }
  handleInput(e) {
    let field = e.target.getAttribute('data-field')
    let value = e.target.value
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        [field]: value
      }
      return newState;
    })
  } 
  handleInputFile(e) {
    let field = e.target.getAttribute('data-field')
    let file = e.target.files
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        [field]: file
      }
      return newState;
    })
  }
  addAlbum() {
    let name = this.state.playlist__name
    let category = this.state.playlist__category
    let desc = this.state.playlist__desc
    let upload = this.state.photo__file[0]
    let storageRef = firebase.storage().ref('photo/' + name)
    let photoUrl = ''
    db.collection('albums').doc(name).get().then((doc) => {
      if (doc.exists) {
        alert('此專輯已存在!')
      }
      else {
        storageRef.put(upload).then(() => {
          storageRef.getDownloadURL().then(function (photo__url) {
            photoUrl = photo__url
            db.collection('albums').doc(name).set({
              name,
              category,
              desc,
              photoUrl,
              songs: []
            }).then(() => {
              alert('新增成功!')
            })
          })
        })  
      }
    })
  }
  addSong() {
    let select = this.state.select__playlist
    let album = this.state.song__album
    let name = this.state.song__name
    let singer = this.state.song__singer
    let length = this.state.song__length
    let upload = this.state.song__file[0]
    let storageRef = firebase.storage().ref('mp3/' + name)
    let songUrl = ''
    storageRef.put(upload).then(() => {
      storageRef.getDownloadURL().then(function (song__url) {
        songUrl = song__url
        db.collection('albums').doc(select).get().then((doc) => {
          if (doc.exists) {
            let { songs } = doc.data()
            let newSong = {
              name,
              singer,
              length,
              songUrl,
              album,
            }
            console.log(newSong)
            songs.push(newSong)
            db.collection('albums').doc(select).update({
              songs,
            }).then(() => {
              alert('添加成功!')
            })
          }
        })
      })
    })
  }
}

export default Admin
