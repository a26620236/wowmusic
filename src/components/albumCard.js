import React from 'react';
import ReactDOM from 'react-dom'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";
import { db, firebase } from '../static/js/firebase'

class AlbumCard extends React.Component {
  constructor(props) {
    super(props)
  }
  render() {
    return (
      <Link className='itembox' to={'/album/' + this.props.data.name}>
        <div className='album-img'>
          <img src={this.props.data.photoUrl} />
        </div>
        <div className='headline'>{this.props.data.name}</div>
        <div className='inform'>{this.props.data.desc}</div>
      </Link>
    )
  }
}

export default AlbumCard
