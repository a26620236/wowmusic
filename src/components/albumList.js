import React from 'react';
import ReactDOM from 'react-dom'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";
import AlbumCard from './albumCard'
import { db, firebase } from '../static/js/firebase'

class Albumlist extends React.Component {
  constructor(props) {
    super(props)
  }
  render() {
    let { data } = this.props
    let outer = []
    let inner = []
    let arr = []
    let arr1 =[]
    if (data.length > 0) {
      for (let i=0; i<data.length; i++) {
        for (let j = 0; j < data[i].length; j++) {
          arr1 = <div className='items' key={j}>
            <AlbumCard data={data[i][j]} />
          </div>
          inner.push(arr1)
        }
        arr = <div className='list' key={i}>
          <div className='title'>{data[i][0].category}</div>
          <div className='itemlist'>{inner}</div>
        </div>
        outer.push(arr)
        inner = []
      }
      return (
        <div>
          {outer}
        </div>
      )
    }
    else {
      return (
        <div>isLoading</div>
      )
    }
  }
}

export default Albumlist
