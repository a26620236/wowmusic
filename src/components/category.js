import React from 'react';
import ReactDOM from 'react-dom'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";
import Header from './headerNormal'
import AlbumCard from './albumCard'
import { db, firebase, storage } from '../static/js/firebase'

class Category extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      data: [],
    }
  }
  componentDidMount() {
    let data = []
    let { category } = this.props
    db.collection("albums").where('category', '==', category).get().then((querySnapshot) => {
      querySnapshot.forEach(function (doc) {
        data.push(doc.data())
      });
      this.setState((currentState) => {
        let newState = {
          ...currentState,
          data,
        }
        return newState
      })
    })
  }
  render() {
    let { data } = this.state
    let { isLogin, isAdmin, user , category } = this.props
    if (data.length > 0) {
      return (
        <div className='category'>
          <Header isLogin={isLogin} isAdmin={isAdmin} user={user}/>
          <div className='category-wrapper'>
            <div className='banner'>
              {category}
            </div>
            <div className='wrapper__background'>
              <div className='title'>熱門播放清單</div>
              <div className='items'>
                {data.map((e, index) => {
                  return <AlbumCard data={e} key={index} />
                })}
              </div>
            </div>
          </div>
        </div>
      )
    }
    else {
      return (
        <div className='isloading'>
          <Header isLogin={isLogin} isAdmin={isAdmin} user={user} />
        </div>
      )
    }
  }
}

export default Category
