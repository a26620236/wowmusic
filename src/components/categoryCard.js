import React from 'react';
import ReactDOM from 'react-dom'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";

class CategoryCard extends React.Component {
  constructor(props) {
    super(props)
  }
  render() {
    return (
      <div className='itembox'>
        <div className='category'>居家生活</div>
      </div>
    )
  }
}

export default CategoryCard
