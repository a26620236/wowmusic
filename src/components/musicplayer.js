import React from 'react';
import ReactDOM from 'react-dom'
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link
} from "react-router-dom";
import { db, firebase, storage } from '../static/js/firebase'

class MusicPlayer extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      play: true,
      currentTime: 0,
      playMode: 'normal',
      randomIndex: 0,
      mobilePlayerState: false,
      mobileBarState: true,
    }
    this.audio = React.createRef()
    this.timeline = React.createRef()
    this.mobile__timeline = React.createRef()
    this.handle = React.createRef()
    this.mobile__handle = React.createRef()
    this.volume = React.createRef()
    this.mobile__volume = React.createRef()
    this.handle__volume = React.createRef()
    this.mobile__handle__volume = React.createRef()
    this.mouseUp = this.mouseUp.bind(this)
    this.mouseMove = this.mouseMove.bind(this)
    this.mouseDown = this.mouseDown.bind(this)
    this.volumeMove = this.volumeMove.bind(this)
    this.mouseUpVolume = this.mouseUpVolume.bind(this)
    this.mouseDownVolume = this.mouseDownVolume.bind(this)
    //mobile
    this.mobileMouseUp = this.mobileMouseUp.bind(this)
    this.mobileMouseMove = this.mobileMouseMove.bind(this)
    this.mobileMouseDown = this.mobileMouseDown.bind(this)
    this.mobileVolumeMove = this.mobileVolumeMove.bind(this)
    this.mobileMouseUpVolume = this.mobileMouseUpVolume.bind(this)
    this.mobileMouseDownVolume = this.mobileMouseDownVolume.bind(this)
    this.play = this.play.bind(this)
    this.setRamdomState = this.setRamdomState.bind(this)
    this.setNormalState = this.setNormalState.bind(this)
  }
  componentDidMount() {
    let audio = this.audio.current
    let timeline = this.timeline.current
    let mobile__timeline = this.mobile__timeline.current
    let timelineLeft = timeline.getBoundingClientRect()
    let mobile__timelineLeft = mobile__timeline.getBoundingClientRect()
    audio.addEventListener("timeupdate", () => {
      let ratio = audio.currentTime / audio.duration;
      let position = (timeline.offsetWidth * ratio) + timelineLeft.left;

      this.positionHandle(position);
      this.setState((currentState) => {
        let newState = {
          ...currentState,
          currentTime: Math.floor(audio.currentTime),
        }
        return newState
      })
    })
    audio.addEventListener("timeupdate", () => {
      let ratio = audio.currentTime / audio.duration;
      let position = (mobile__timeline.offsetWidth * ratio) + mobile__timelineLeft.left;

      this.mobilePositionHandle(position);
    });
    audio.addEventListener("ended", () => {
      let { songs, playIndex } = this.props.playlist[0]
      let { playLoop } = this.props
      let { playMode } = this.state

      if (playIndex == (songs.length - 1) && playMode === 'normal') {
        this.setState((currentState) => {
          let newState = {
            ...currentState,
            play: false
          }
          return newState
        })
      }
      if (playIndex < (songs.length - 1) && playMode === 'normal') {
        this.clickNext()
      }
      if (playIndex < (songs.length - 1) && playMode === 'loop') {
        this.clickNext()
        return
      }
      if (playIndex == (songs.length - 1) && playMode === 'loop') {
        playLoop()
        return
      }
    });
  }
  componentDidUpdate(prevProps) {
    if (this.props !== prevProps) {
      this.setState((currentState) => {
        let newState = {
          ...currentState,
          play: true
        }
        return newState
      })
    }
  }
  render() {
    let audio = this.audio.current
    if (audio == null) {
      return (
        <div className='musicplayer'> 
          <audio src='' ref={this.audio} />
          <div className='player-container'>
            <div className='mobile-bar'>
              <div className='show' onClick={this.clickMobilebar.bind(this)}><i className="fas fa-angle-left"></i></div>
              <div className='playing'><i className="fas fa-music"></i></div>
              <div className='hide' onClick={this.clickMobilebar.bind(this)}><i className="fas fa-angle-right"></i></div>
            </div>
            <div className='now-playing-bar'>
              <div className='now-playing-bar__left'>
                <div className='container'>
                  <div className='album'>
                    <img src="https://i.scdn.co/image/ab67616d000048513d0ffcf6ee624625158fa352" />
                  </div>
                  <div className='song-inform'>
                    <div className='name'>太陽</div>
                    <div className='singer'>邱振哲</div>
                  </div>
                  <div className='like-btn'>
                    <i className="far fa-heart"></i>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__center'>
                <div className='container'>
                  <div className='player-controls__btns'>
                    <div><i className="fas fa-random"></i></div>
                    <div><i className="fas fa-step-backward"></i></div>
                    <div onClick={this.play}><i className={!this.state.play ? "far fa-play-circle" : "far fa-pause-circle"}></i></div>
                    <div onClick={this.clickNext.bind(this)}><i className="fas fa-step-forward"></i></div>
                    <div><i className="fas fa-retweet"></i></div>
                  </div>
                  <div className='playback-bar'>
                    <div className='progress-time'>0:00</div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='timeline' onClick={this.mouseMove} ref={this.timeline}>
                          <div className='handle' onMouseDown={this.mouseDown} ref={this.handle}></div>
                        </div>
                      </div>
                    </div>
                    <div className='song-length'>4:22</div>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__right'>
                <div className='container'>
                  <div className='waiting-list'>
                    <Link to='/musicqueue'>
                      <i className="fas fa-bars"></i>
                    </Link>
                  </div>
                  <div className='volume-bar'>
                    <div className='volume-btn'>
                      <i className="fas fa-volume-up"></i>
                    </div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='volume' onClick={this.volumeMove} ref={this.volume}>
                          <div className='handle__volume' onMouseDown={this.mouseDownVolume} ref={this.handle__volume}></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div> 
          <div className='mobile-player'>
            <img src='' className='background'></img>
            <div className='now-playing-bar'>
              <div className='now-playing-bar__left'>
                <div className='container'>
                  <div className='album'>
                    <img src='' />
                  </div>
                  <div className='song-inform'>
                    <div className='name'></div>
                    <div className='singer'></div>
                  </div>
                  <div className='like-btn'>
                    <i className="far fa-heart"></i>
                    <div className='waiting-list' onClick={this.clickMobilePlayer.bind(this)}>
                      <Link to='/musicqueue'>
                        <i className="fas fa-bars"></i>
                      </Link>
                    </div>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__center'>
                <div className='container'>
                  <div className='player-controls__btns'>
                    <div onClick={this.clickRamdom.bind(this)}><i className="fas fa-random"></i></div>
                    <div onClick={this.clickPrevious.bind(this)}><i className="fas fa-step-backward"></i></div>
                    <div className='play-btn' onClick={this.play}><i className={!this.state.play ? "far fa-play-circle" : "far fa-pause-circle"}></i></div>
                    <div onClick={this.clickNext.bind(this)}><i className="fas fa-step-forward"></i></div>
                    <div onClick={this.clickLoop.bind(this)}><i className="fas fa-retweet"></i></div>
                  </div>
                  <div className='playback-bar'>
                    <div className='progress-time'></div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='timeline' onClick={this.mobileMouseMove} ref={this.mobile__timeline}>
                          <div className='handle' onMouseDown={this.mobileMouseDown} ref={this.mobile__handle}></div>
                        </div>
                      </div>
                    </div>
                    <div className='song-length'></div>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__right'>
                <div className='container'>
                  <div className='waiting-list'>
                    <Link to='/musicqueue'>
                      <i className="fas fa-bars"></i>
                    </Link>
                  </div>
                  <div className='volume-bar'>
                    <div className='volume-btn'>
                      <i className="fas fa-volume-up"></i>
                    </div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='volume' onClick={this.mobileVolumeMove} ref={this.mobile__volume}>
                          <div className='handle__volume' onMouseDown={this.mobileMouseDownVolume} ref={this.mobile__handle__volume}></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div className='mobile-close-btn' onClick={this.clickMobilePlayer.bind(this)}><i className="fas fa-chevron-down"></i></div>
            </div>
          </div>
        </div>
      )
    }
    else {
      let { currentTime, playMode, randomIndex, mobilePlayerState, mobileBarState } = this.state
      let { playlist } = this.props
      let { songs, playIndex, photoUrl } = playlist[0]

      if (currentTime === 0) {
        currentTime = '0:00'
      }
      else if (currentTime < 10) {
        currentTime = '0:0' + currentTime
      }
      else if (currentTime >= 10 && currentTime < 60) {
        currentTime = '0:' + currentTime
      }
      else {
        let seconds = Math.floor(currentTime % 60)
        if (seconds < 10) seconds = '0' + seconds
        currentTime = Math.floor(currentTime / 60).toString() + ':' + seconds
      }
      return (
        <div className='musicplayer'>
          <audio src={playMode === 'normal' || playMode === 'loop' ? songs[playIndex].songUrl : songs[randomIndex].songUrl} ref={this.audio} autoPlay={true} />
          <div className={mobileBarState ? 'player-container' : 'player-container-closed'}>
            <div className='mobile-bar'>
              <div className='show' onClick={this.clickMobilebar.bind(this)}><i className="fas fa-angle-left"></i></div>
              <div className='playing'><i className="fas fa-music"></i></div>
              <div className='hide' onClick={this.clickMobilebar.bind(this)}><i className="fas fa-angle-right"></i></div>
            </div>
            <div className='now-playing-bar'>
              <div className='now-playing-bar__left' onClick={this.clickMobilePlayer.bind(this)}>
                <div className='container'>
                  <div className='album'>
                    <img src={photoUrl} />
                  </div>
                  <div className='song-inform'>
                    <div className='name'>{songs[playIndex].name}</div>
                    <div className='singer'>{songs[playIndex].singer}</div>
                  </div>
                  <div className='like-btn'>
                    <i className="far fa-heart"></i>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__center'>
                <div className='container'>
                  <div className='player-controls__btns'>
                    <div onClick={this.clickRamdom.bind(this)}><i className="fas fa-random"></i></div>
                    <div onClick={this.clickPrevious.bind(this)}><i className="fas fa-step-backward"></i></div>
                    <div className='play-btn' onClick={this.play}><i className={!this.state.play ? "far fa-play-circle" : "far fa-pause-circle"}></i></div>
                    <div onClick={this.clickNext.bind(this)}><i className="fas fa-step-forward"></i></div>
                    <div onClick={this.clickLoop.bind(this)}><i className="fas fa-retweet"></i></div>
                  </div>
                  <div className='playback-bar'>
                    <div className='progress-time'>{currentTime}</div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='timeline' onClick={this.mouseMove} ref={this.timeline}>
                          <div className='handle' onMouseDown={this.mouseDown} ref={this.handle}></div>
                        </div>
                      </div>
                    </div>
                    <div className='song-length'>{songs[playIndex].length}</div>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__right'>
                <div className='container'>
                  <div className='waiting-list'>
                    <Link to='/musicqueue'>
                      <i className="fas fa-bars"></i>
                    </Link>
                  </div>
                  <div className='volume-bar'>
                    <div className='volume-btn'>
                      <i className="fas fa-volume-up"></i>
                    </div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='volume' onClick={this.volumeMove} ref={this.volume}>
                          <div className='handle__volume' onMouseDown={this.mouseDownVolume} ref={this.handle__volume}></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div className={mobilePlayerState ? 'mobile-player' : 'mobile-player-closed'}>
            <img src={photoUrl} className='background'></img>
            <div className='now-playing-bar'>
              <div className='now-playing-bar__left'>
                <div className='container'>
                  <div className='album'>
                    <img src={photoUrl} />
                  </div>
                  <div className='song-inform'>
                    <div className='name'>{songs[playIndex].name}</div>
                    <div className='singer'>{songs[playIndex].singer}</div>
                  </div>
                  <div className='like-btn'>
                    <i className="far fa-heart"></i>
                    <div className='waiting-list' onClick={this.clickMobilePlayer.bind(this)}>
                      <Link to='/musicqueue'>
                        <i className="fas fa-bars"></i>
                      </Link>
                    </div>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__center'>
                <div className='container'>
                  <div className='player-controls__btns'>
                    <div onClick={this.clickRamdom.bind(this)}><i className="fas fa-random"></i></div>
                    <div onClick={this.clickPrevious.bind(this)}><i className="fas fa-step-backward"></i></div>
                    <div className='play-btn' onClick={this.play}><i className={!this.state.play ? "far fa-play-circle" : "far fa-pause-circle"}></i></div>
                    <div onClick={this.clickNext.bind(this)}><i className="fas fa-step-forward"></i></div>
                    <div onClick={this.clickLoop.bind(this)}><i className="fas fa-retweet"></i></div>
                  </div>
                  <div className='playback-bar'>
                    <div className='progress-time'>{currentTime}</div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='timeline' onClick={this.mobileMouseMove} ref={this.mobile__timeline}>
                          <div className='handle' onMouseDown={this.mobileMouseDown} ref={this.mobile__handle}></div>
                        </div>
                      </div>
                    </div>
                    <div className='song-length'>{songs[playIndex].length}</div>
                  </div>
                </div>
              </div>
              <div className='now-playing-bar__right'>
                <div className='container'>
                  <div className='waiting-list' onClick={this.clickMobilePlayer.bind(this)}>
                    <Link to='/musicqueue'>
                      <i className="fas fa-bars"></i>
                    </Link>
                  </div>
                  <div className='volume-bar'>
                    <div className='volume-btn'>
                      <i className="fas fa-volume-up"></i>
                    </div>
                    <div className='progress-bar'>
                      <div className='container'>
                        <div className='volume' onClick={this.mobileVolumeMove} ref={this.mobile__volume}>
                          <div className='handle__volume' onMouseDown={this.mobileMouseDownVolume} ref={this.mobile__handle__volume}></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div className='mobile-close-btn' onClick={this.clickMobilePlayer.bind(this)}><i className="fas fa-chevron-down"></i></div>
            </div>
          </div>
        </div>
      )
    }
  }
  play() {
    let audio = this.audio.current
    if (this.state.play) {
      this.setState((currentState) => {
        let newState = {
          ...currentState,
          play: false
        }
        return newState
      })
      audio.pause()
    } else {
      this.setState((currentState) => {
        let newState = {
          ...currentState,
          play: true
        }
        return newState
      })
      audio.play()
    }
  }
  positionHandle(position) {
    let timeline = this.timeline.current
    let timelineLeft = timeline.getBoundingClientRect()
    let handle = this.handle.current
    let timelineWidth = timeline.offsetWidth - handle.offsetWidth
    let handleLeft = position - timelineLeft.left
    
    if (handleLeft > 0 && handleLeft <= timelineWidth) {
      handle.style.marginLeft = handleLeft + 'px'
    }
    if (handleLeft < 0) {
      handle.style.marginLeft = '0px'
    }
    if (handleLeft > timelineWidth) {
      handle.style.marginLeft = timelineWidth + 'px'
    }
  }
  volumeHandle(position) {
    let volume = this.volume.current
    let volumeLeft = volume.getBoundingClientRect()
    let handle__volume = this.handle__volume.current
    let volumeWidth = volume.offsetWidth - handle__volume.offsetWidth
    let handle__volumeLeft = position - volumeLeft.left
    
    if (handle__volumeLeft > 0 && handle__volumeLeft <= volumeWidth) {
      handle__volume.style.marginRight = (volumeWidth - handle__volumeLeft + 12) + 'px'
    }
    if (handle__volumeLeft < 0) {
      handle__volume.style.marginRight = volumeWidth + 'px'
    }
    if (handle__volumeLeft > volumeWidth) {
      handle__volume.style.marginRight = '0px'
    }
  }
  volumeMove(e) {
    let audio = this.audio.current
    let volume = this.volume.current
    let volumeLeft = volume.getBoundingClientRect()

    this.volumeHandle(e.pageX)
    let sound = ((e.pageX - volumeLeft.left) / volume.offsetWidth).toFixed(1)
    if (sound <= 1) {
      audio.volume = sound
    }
    else {
      audio.volume = 1
    }
  }
  mouseMove(e) {
    let audio = this.audio.current
    let timeline = this.timeline.current
    let timelineLeft = timeline.getBoundingClientRect()
    
    this.positionHandle(e.pageX)
    audio.currentTime = ((e.pageX - timelineLeft.left) / timeline.offsetWidth) * audio.duration
  }
  mouseUp(e) {
    window.removeEventListener('mousemove', this.mouseMove);
    window.removeEventListener('mouseup', this.mouseUp);
  };
  mouseDown(e) {
    window.addEventListener('mousemove', this.mouseMove);
    window.addEventListener('mouseup', this.mouseUp);
  };
  mouseUpVolume(e) {
    window.removeEventListener('mousemove', this.volumeMove);
    window.removeEventListener('mouseup', this.mouseUpVolume);
  }
  mouseDownVolume(e) {
    window.addEventListener('mousemove', this.volumeMove);
    window.addEventListener('mouseup', this.mouseUpVolume);
  }
  // mobile
  mobilePositionHandle(position) {
    let mobile__timeline = this.mobile__timeline.current
    let mobile__timelineLeft = mobile__timeline.getBoundingClientRect()
    let mobile__handle = this.mobile__handle.current
    let mobile__timelineWidth = mobile__timeline.offsetWidth - mobile__handle.offsetWidth
    let mobile__handleLeft = position - mobile__timelineLeft.left

    if (mobile__handleLeft > 0 && mobile__handleLeft <= mobile__timelineWidth) {
      mobile__handle.style.marginLeft = mobile__handleLeft + 'px'
    }
    if (mobile__handleLeft < 0) {
      mobile__handle.style.marginLeft = '0px'
    }
    if (mobile__handleLeft > mobile__timelineWidth) {
      mobile__handle.style.marginLeft = mobile__timelineWidth + 'px'
    }
  }
  mobileVolumeHandle(position) {
    let mobile__volume = this.mobile__volume.current
    let mobile__volumeLeft = mobile__volume.getBoundingClientRect()
    let mobile__handle__volume = this.mobile__handle__volume.current
    let volumeWidth = mobile__volume.offsetWidth - mobile__handle__volume.offsetWidth
    let mobile__handle__volumeLeft = position - mobile__volumeLeft.left

    if (mobile__handle__volumeLeft > 0 && mobile__handle__volumeLeft <= volumeWidth) {
      mobile__handle__volume.style.marginRight = (volumeWidth - mobile__handle__volumeLeft + 12) + 'px'
    }
    if (mobile__handle__volumeLeft < 0) {
      mobile__handle__volume.style.marginRight = volumeWidth + 'px'
    }
    if (mobile__handle__volumeLeft > volumeWidth) {
      mobile__handle__volume.style.marginRight = '0px'
    }
  }
  mobileVolumeMove(e) {
    let audio = this.audio.current
    let mobile__volume = this.mobile__volume.current
    let mobile__volumeLeft = mobile__volume.getBoundingClientRect()

    this.mobileVolumeHandle(e.pageX)
    let sound = ((e.pageX - mobile__volumeLeft.left) / mobile__volume.offsetWidth).toFixed(1)
    if (sound <= 1) {
      audio.volume = sound
    }
    else {
      audio.volume = 1
    }
  }
  mobileMouseMove(e) {
    let audio = this.audio.current
    let mobile__timeline = this.mobile__timeline.current
    let mobile__timelineLeft = mobile__timeline.getBoundingClientRect()

    this.mobilePositionHandle(e.pageX)
    audio.currentTime = ((e.pageX - mobile__timelineLeft.left) / mobile__timeline.offsetWidth) * audio.duration
  }
  mobileMouseUp(e) {
    window.removeEventListener('mousemove', this.mobileMouseMove);
    window.removeEventListener('mouseup', this.mobileMouseUp);
  };
  mobileMouseDown(e) {
    window.addEventListener('mousemove', this.mobileMouseMove);
    window.addEventListener('mouseup', this.mobileMouseUp);
  };
  mobileMouseUpVolume(e) {
    window.removeEventListener('mousemove', this.mobileVolumeMove);
    window.removeEventListener('mouseup', this.mobileMouseUpVolume);
  }
  mobileMouseDownVolume(e) {
    window.addEventListener('mousemove', this.mobileVolumeMove);
    window.addEventListener('mouseup', this.mobileMouseUpVolume);
  }
  clickNext() {
    let { playNext } = this.props
    let { playMode } = this.state
    let { songs, playIndex } = this.props.playlist[0]
    if (playMode === 'normal' && playIndex < (songs.length - 1)) {
      playNext()
    }
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        play: true
      }
      return newState
    })
  }
  clickPrevious() {
    let { playPrevious } = this.props
    let { playMode } = this.state
    let { playIndex } = this.props.playlist[0]
    if (playMode === 'normal' && playIndex > 0) {
      playPrevious()
    }
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        play: true
      }
      return newState
    })
  }
  clickLoop() {
    let { playMode } = this.state
    let audio = this.audio.current
    if (playMode !== 'loop') {
      audio.removeEventListener("ended", this.setNormalState)
      audio.removeEventListener("ended", this.setRamdomState)
      this.setState((currentState) => {
        let newState = {
          ...currentState,
          playMode: 'loop',
          randomIndex: 0
        }
        return newState
      })
    }
    if (playMode === 'loop') {
      audio.addEventListener("ended", this.setNormalState)
    }
  }
  clickRamdom() {
    let { playMode } = this.state
    let audio = this.audio.current

    if (playMode !== 'random') {
      audio.removeEventListener("ended", this.setNormalState)
      audio.addEventListener("ended", this.setRamdomState)
    }
    if (playMode === 'random') {
      audio.removeEventListener("ended", this.setRamdomState)
      audio.addEventListener("ended", this.setNormalState)
    }
  }
  setRamdomState() {
    let { songs } = this.props.playlist[0]
    let { randomIndex } = this.state
    let setIndex = Math.floor(Math.random() * songs.length)

    while (setIndex == randomIndex) {
      setIndex = Math.floor(Math.random() * songs.length)
    }
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        playMode: 'random',
        randomIndex: setIndex,
        play: true,
      }
      return newState
    })
  }
  setNormalState() {
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        playMode: 'normal',
        randomIndex: 0
      }
      return newState
    })
  }
  clickMobilePlayer() {
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        mobilePlayerState: !currentState.mobilePlayerState
      }
      return newState
    })
  }
  clickMobilebar() {
    this.setState((currentState) => {
      let newState = {
        ...currentState,
        mobileBarState: !currentState.mobileBarState
      }
      return newState
    })
  }
}

export default MusicPlayer
