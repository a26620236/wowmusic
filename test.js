import { timeFormat, MobileBar } from './src/components/musicplayer'
import React from 'react';
import renderer from 'react-test-renderer';
import axios from 'axios'

test("test currentTime format", () => {
  expect(timeFormat(25)).toBe('0:25')
  expect(timeFormat(600)).toBe('10:00')
  expect(timeFormat(0)).toBe('0:00')
  expect(timeFormat(null)).toBe('0:00')
  expect(timeFormat(87)).toBe('1:27')
  expect(timeFormat('')).toBe('0:00')
  expect(timeFormat(false)).toBe('0:00')
  expect(timeFormat(true)).toBe('0:00')
})

it('renders correctly', () => {
  let testData = {
    function: {
      clickMobilebar: 'click'
    }
  }
  const tree = renderer
    .create(<MobileBar data={testData}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

const API = 'https://jsonplaceholder.typicode.com/posts/1'

const getData = function() {
  return axios.get(API).then(resp => resp.data)
}

jest.mock('axios');

test('should fetch post', () => {
  const users = {
    "userId": 1,
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "body": "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
  }
  const resp = { data: users }
  axios.get.mockResolvedValue(resp);

  return getData().then(data => expect(data).toEqual(users));
})
