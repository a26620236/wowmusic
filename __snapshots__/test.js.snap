// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="mobile-bar"
>
  <div
    className="show"
    onClick="click"
  >
    <i
      className="fas fa-angle-left"
    />
  </div>
  <div
    className="playing"
  >
    <i
      className="fas fa-music"
    />
  </div>
  <div
    className="hide"
    onClick="click"
  >
    <i
      className="fas fa-angle-right"
    />
  </div>
</div>
`;
